source/common/config/subscription_factory.h

```diff
@@ -0,0 +1,79 @@
+#pragma once
+
+#include "envoy/config/subscription.h"
+
+#include "common/config/filesystem_subscription_impl.h"
+#include "common/config/grpc_subscription_impl.h"
+#include "common/config/http_subscription_impl.h"
+#include "common/config/utility.h"
+
+#include "api/base.pb.h"
+
+namespace Envoy {
+namespace Config {
+
+class SubscriptionFactory {
+public:
+  /**
+   * Subscription factory.
+   * @param config envoy::api::v2::ConfigSource to construct from.
+   * @param node envoy::api::v2::Node identifier.
+   * @param dispatcher event dispatcher.
+   * @param cm cluster manager for async clients (when REST/gRPC).
+   * @param random random generator for jittering polling delays (when REST).
+   * @param rest_legacy_constructor constructor function for Subscription adapters (when legacy v1
+   * REST).
+   * @param rest_method fully qualified name of v2 REST API method (as per protobuf service
+   *        description).
+   * @param grpc_method fully qualified name of v2 gRPC API bidi streaming method (as per protobuf
+   *        service description).
+   */
+  template <class ResourceType>
+  static std::unique_ptr<Subscription<ResourceType>>
+  subscriptionFromConfigSource(const envoy::api::v2::ConfigSource& config,
+                               const envoy::api::v2::Node& node, Event::Dispatcher& dispatcher,
+                               Upstream::ClusterManager& cm, Runtime::RandomGenerator& random,
+                               std::function<Subscription<ResourceType>*()> rest_legacy_constructor,
+                               const std::string& rest_method, const std::string& grpc_method) {
+    std::unique_ptr<Subscription<ResourceType>> result;
+    switch (config.config_source_specifier_case()) {
+    case envoy::api::v2::ConfigSource::kPath:
+      result.reset(new Config::FilesystemSubscriptionImpl<ResourceType>(dispatcher, config.path()));
```

Comment:
super nit: You can actually just directly return the unique pointer by doing something like:
`return std::unique_ptr<Subscription<ResourceType>>{new ...}`. This will remove a lot of break statements, etc. Can also be made simpler probably with a local typedef. Doesn't matter if you change this or not just pointing it out.